### ISSUE TITLE
Proposal: [API] Inspection aggregation

### ISSUE THREAD

**Issue by MarcoLotz** [docker/docker#18162]
There are some information that can only be reliably retrieved using the container inspection.

When you get normal container information (e.g. through container listing), one cannot find out in what network is a container in. Before api 1.21, one could not find the imageID of the containers without inspecting it too - sometimes it returned the image repo+tag and sometimes the id.

When a single host has thousands of containers, getting network information for every single container is very expensive, since it demands a get request per container.

I would like to suggest an API extension where all the containers' inspection information will be retrieved in a single request. Although the response may be considerable large (several jsons), it is still better than a restful request per inspection.

**Comment by LK4D4** [comment]
There is still no batch requests. I'm not sure that this is still an issue with keep-alive though, headers overhead is not so big.
WDYT @dnephin @thaJeztah @tonistiigi

**Comment by dnephin** [comment]
I think if we were to make this change we would have to do it for every `GET` and `DELETE` endpoint. It's not a small change, so there needs to be significant benefit. In most cases just doing multiple calls is still pretty fast, especially from the same host.

Related to #5893, it might make sense as part of a larger API change/cleanup

**Comment by MarcoLotz** [comment]
@dnephin @LK4D4 there are very good reasons for that. Retrieving complete container information for large machines (e.g. HPE DL980 - 4TB Ram, 80 cores; IBM POWER8 - up to 56 TB of extended memory) is really expensive in current API. This kind of machines may be apple to have thousands of containers running at a single piece of metal. In the past, I've depployed 10k containers on one of them.

With the current API, that resulted in 10k Restful requests per update cycle if you want all the container inspection. There are many information about the container that can only be retrieved through inspection - and thus is not available by only listing. 

Basically, the benefit depends if docker wants to be scale for large memory machines or not. One may say that using a cache may be a good solution, but is quite expensive for dynamic container allocation environments - which is a really common environment.
